src/util-ja3.c

```diff
@@ -0,0 +1,290 @@
+
+    return 0;
+}
+
+/**
+ * \internal
+ * \brief Return number of digits in number.
+ *
+ * \param num The number.
+ *
+ * \return digits Number of digits.
+ */
+static unsigned NumberOfDigits(unsigned num)
+{
+    if (num < 10) {
+        return 1;
+    }
+
+    return 1 + NumberOfDigits(num / 10);
+}
+
+/**
+ * \brief Add value to buffer.
+ *
+ * \param buffer The buffer.
+ * \param value  The value.
+ *
+ * \retval 0 on success.
+ * \retval -1 on failure.
+ */
+int Ja3BufferAddValue(JA3Buffer *buffer, int value)
+{
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be NULL");
+        return -1;
+    }
+
+    if (buffer->data == NULL) {
+        buffer->data = SCMalloc(JA3_BUFFER_INITIAL_SIZE * sizeof(char));
+        if (buffer->data == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC,
+                       "Error allocating memory for JA3 data");
+            Ja3BufferFree(buffer);
+            return -1;
+        }
+        buffer->size = JA3_BUFFER_INITIAL_SIZE;
+    }
+
+    unsigned value_len = NumberOfDigits(value);
+
+    int rc = Ja3BufferResizeIfFull(buffer, value_len);
+    if (rc != 0) {
+        Ja3BufferFree(buffer);
+        return -1;
+    }
+
+    if (buffer->used == 0) {
+        buffer->used += snprintf(buffer->data, buffer->size, "%d", value);
+    }
+    else {
+        buffer->used += snprintf(buffer->data + buffer->used, buffer->size -
+                                 buffer->used, "-%d", value);
+    }
+
+    return 0;
+}
+
+/**
+ * \brief Generate Ja3 hash string.
+ *
+ * \param buffer The Ja3 buffer.
+ *
+ * \retval pointer to hash string on success.
+ * \retval NULL on failure.
+ */
+char *Ja3GenerateHash(JA3Buffer *buffer)
+{
+
+#ifdef HAVE_NSS
+    if (buffer == NULL) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "Buffer should not be NULL");
+        return NULL;
+    }
+
+    if (buffer->data == NULL) {
+        SCLogError(SC_ERR_INVALID_VALUE, "Buffer data should not be NULL");
+        return NULL;
+    }
+
+    char *ja3_hash = SCMalloc(MD5_STRING_LENGTH * sizeof(char));
+    if (ja3_hash == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC,
+                   "Error allocating memory for JA3 hash");
+        return NULL;
+    }
+
+    unsigned char md5[MD5_LENGTH];
+    HASH_HashBuf(HASH_AlgMD5, md5, (unsigned char *)buffer->data, buffer->used);
+
+    int i, x;
+    for (i = 0, x = 0; x < MD5_LENGTH; x++) {
+        i += snprintf(ja3_hash + i, MD5_STRING_LENGTH - i, "%02x", md5[x]);
+    }
+
+    return ja3_hash;
+#else
+    return NULL;
+#endif /* HAVE_NSS */
+
+}
+
+/**
+ * \brief Check if JA3 is disabled.
+ *
+ * Issue warning if JA3 is disabled or if we are lacking support for JA3.
+ *
+ * \param type Type to add to warning.
+ *
+ * \retval 1 if disabled.
+ * \retval 0 otherwise.
+ */
+int Ja3IsDisabled(const char *type)
+{
+    int is_enabled = 0;
+
+    /* Check if JA3 is enabled */
+    ConfGetBool("app-layer.protocols.tls.ja3-fingerprints", &is_enabled);
```

Comment:
As per our discussion with  @thus  - we ought to make that configure option more visible - aka add it in yaml/doc

Reply:
I'll add a notice in the docs and add it to the yaml.